### DISCUSSION TITLE
How can I convert Tensor from/to numpy array?

### DISCUSSION THREAD

**Discussion by pierric** [modular/modular#1048] [Q&A]
I created a Tensor object, and applied some operations. but now I don't know how can I view the tensor? or if possible can I convert it to numpy array so that I can apply some python function?

**Comment by ajayspatil7** [comment]
Import the tensorprint from tensorutil to print the tensor.

from tensorutils import tensorprint

t = rand[type](2, 3,3)

tensorprint(t)

OR

You can iterate through it and check what's in it.

**Reply by adakoda** [reply]
I have the same question.

The official example used the tensor to numpy code in an inefficient way, but I am looking for a more efficient way.
- [Mandelbrot in Mojo with Python plots](https://docs.modular.com/mojo/notebooks/Mandelbrot.html)

```
def show_plot(tensor: Tensor[float_type]):
    ...
    np = Python.import_module("numpy")
    ...
    for row in range(height):
        for col in range(width):
            numpy_array.itemset((col, row), tensor[col, row])
    ...
```

**Comment by ajayspatil7** [comment]
> Import the tensorprint from tensorutil to print the tensor.
> 
> from tensorutils import tensorprint
> 
> t = rand[type](2, 3,3)
> 
> tensorprint(t)
> 
> OR
> 
> You can iterate through it and check what's in it.

**Comment by marcelroed** [comment]
Here's an example of how to interpret a Mojo `Tensor` as a NumPy array:

```mojo
def func(tensor:​ Tensor[DType.int32]​)
    np = Python.import_module("numpy")
    ctypeslib = Python.import_module("numpy.ctypeslib")
    ctypes = Python.import_module("ctypes")

    var rows = tensor.shape()[0]
    var cols = tensor.shape()[1]

    data_ptr = ctypes.cast(int(tensor._ptr), ctypes.POINTER(ctypes.c_int32))
    numpy_array = ctypeslib.as_array(data_ptr, shape=(rows, cols))
    
    # numpy_array now contains the tensor's content, no traversal required!
```

**Reply by alisterburt** [reply]
Thanks for this! Is there similar guidance somewhere on how to turn a numpy array into a Mojo `Tensor`? Everything I saw in the docs was only concerned with Mojo -> Python

**Reply by marcelroed** [reply]
You could do something like this to use the existing data pointer and memory but interpret it as a Mojo Tensor:

```mojo
def np_to_tensor():
    np = Python.import_module('numpy')
    
    l = [1, 2, 3, 4]
    a = np.array(l, dtype=np.int32)
    pointer = int(a.ctypes.data)
    var t = Tensor[DType.int32](TensorShape(4), DTypePointer[DType.int32, 0](address=pointer))

    print(t)
 ```

But this will cause a double free when both the tensor and the numpy array go out of scope. Instead I would allocate a new tensor and copy the numpy array to it in a single memcopy with the following code:

```mojo
from memory.memory import memcpy
def np_to_tensor():
    np = Python.import_module('numpy')
    
    l = [1, 2, 3, 4]
    a = np.array(l, dtype=np.int32)
    pointer = int(a.ctypes.data)

    var t = Tensor[DType.int32](TensorShape(4))
    memcpy(t._ptr, DTypePointer[DType.int32](address=pointer), 4)  # Copy 4 elements from the NumPy array to the Mojo Tensor

    print(t)
```

**Reply by lbartworks** [reply]
I've landed in this page having the same issue of transferring a multidimensional numpy to a tensor, or to mojo in general.

1) The code sent by @marcelroed does the job for that 1D array, but when I try to move it to two dimensions, does not work (prob Im doing something wrong)

```
from python import Python
from tensor import Tensor, TensorShape
from memory import memcpy

def main():
    var np = Python.import_module("numpy")
    var numpy_array :PythonObject
    numpy_array = np.random.rand(3,3)
    print(numpy_array)

    var tensor_array= Tensor[DType.float32](TensorShape(3,3))
    pointer = int(numpy_array.ctypes.data)
    memcpy(tensor_array._ptr, DTypePointer[DType.float32](pointer), 9)

    print(tensor_array)
```

Fails with 
SIToFP result must be FP or FP vector

2) on a broder advise, what would be the more suitable way to carry out this task? I have seen in different parts of the documentation, different approaches to host the mojo native "numpy equivalent". Use a mojo tensor? a custom mojo matrix struct? collapse the multiD numpy in a list and then initalize the mojo structure later?

thanks
luis

**Reply by lbartworks** [reply]
Hi all, 
Answering myself, the best way I found so far (in the lack currently of a full-equivalent of a numpy tensor in mojo), is use or modify the MojoMatrix code you can find here:

https://github.com/modularml/devrel-extras/blob/main/blogs/mojo-row-major-column-major/row_col_matrix.mojo

**Comment by StijnWoestenborghs** [comment]
Basalt has some utilities for that as well:
https://github.com/basalt-org/basalt/blob/main/basalt/utils/tensor_creation_utils.mojo
